# In-Class Coding Lab: Files

The goals of this lab are to help you to understand:

- Reading data from a file all at once or one line at a time.
- Searching for data in files
- Parsing text data to numerical data.
- How to build complex programs incrementally.

## Average  Spam Confidence

For this lab, we will write a program to read spam confidence headers from a mailbox file like `ICCL-mbox-tiny.txt` or `ICCL-mbox-small.txt`. These files contain raw email data, and in that data is a SPAM confidence number for each message:

`X-DSPAM-Confidence:0.8475`

Our goal will be to find each of these lines in the file, and extract the confidence number (In this case `0.8475`), with the end-goal of calculating the average SPAM Confidence of all the emails in the file. 

### Reading from the file

Let's start with some code to read the lines of text from `ICCL-mbox-tiny.txt`

In [13]:
filename = "CCL-mbox-tiny.txt"
count = 0
with open(filename, 'r') as f:
    for line in f.readlines():
        count = count + 1
print(count)

332


### Now Try It

Now modify the code above to print the number of lines in the file, instead of printing the lines themselves. You'll need to increment a variable each time through the loop and then print it out afterwards.

There should be **332** lines.

### Finding the SPAM Confidence

Next, we'll focus on only getting lines addressing `X-DSPAM-Confidence:`. We do this by including an `if` statement inside  the `for` loop.

You need to edit line 4 of the code below to only print lines which begin with `X-DSPAM-Confidence:` There should be **5**

In [15]:
filename = "CCL-mbox-tiny.txt"
count=0
with open(filename, 'r') as f:
    for line in f.readlines():
        if line.startswith("X-DSPAM-Confidence:"):
            count = count + 1
            print(count)


1
2
3
4
5


### Parsing out the confidence value

The final step is to figure out how to parse out the confidence value from the string. 
For example for the given line: `X-DSPAM-Confidence: 0.8475` we need to get the value `0.8475` as a float.

The  strategy here is to replace `X-DSPAM-Confidence:` with an empty string, then calling the `float()` function to convert the results to a float. 

### Now Try It


In [12]:
line = 'X-DSPAM-Confidence: 0.8475'
number =  line.strip("X-DSPAM-Confidence: ")
number = float(number)
print(number)

0.8475


### Putting it all together

Now that we have all the working parts, let's put it all together.

```
1.  line count is 0
2.  total confidence is 0
3.  open mailbox file
4.  for each line in file
5.  if line starts with `X-DSPAM-Confidence:`
6.     remove `X-DSPAM-Confidence:` from line and convert to float
7.     increment line count
8.     add spam confidence to total confidence
9.  print average confidence (total confidence/line count)
```

In [18]:
## TODO: Write program here:
linecount = 0
totalconf = 0
filename = "CCL-mbox-tiny.txt"

with open(filename, 'r') as f:
    for line in f.readlines():
        if line.startswith("X-DSPAM-Confidence:"):
            number = line.strip("X-DSPAM-Confidence: ")
            number = float(number)
            print (number)
            linecount = linecount + 1
            totalconf = totalconf + number
            print(totalconf)
print(totalconf / linecount)
            


0.8475
0.8475
0.6178
1.4653
0.6961
2.1614
0.7565
2.9179
0.7626
3.6805
0.7361


When you have the program working, try it with `ICCL-mbox-short.txt` mailbox file, too.